src/datajson.c

```diff
@@ -0,0 +1,1009 @@
+
+    if (set == NULL)
+        return rrep;
+
+    if (data_len != SC_SHA256_LEN)
+        return rrep;
+
+    Sha256Type lookup = { .json.value = NULL, .json.len = 0 };
+    memcpy(lookup.sha256, data, data_len);
+    THashData *rdata = THashLookupFromHash(set->hash, &lookup);
+    if (rdata) {
+        Sha256Type *found = rdata->data;
+        rrep.found = true;
+        rrep.json = found->json;
+        DatajsonUnlockData(rdata);
+        return rrep;
+    }
+    return rrep;
+}
+
+static DataJsonResultType DatajsonLookupIPv4(
+        Dataset *set, const uint8_t *data, const uint32_t data_len)
+{
+    DataJsonResultType rrep = { .found = false, .json = { .value = NULL, .len = 0 } };
+
+    if (set == NULL)
+        return rrep;
+
+    if (data_len != SC_IPV4_LEN)
+        return rrep;
+
+    IPv4Type lookup = { .json.value = NULL, .json.len = 0 };
+    memcpy(lookup.ipv4, data, data_len);
+    THashData *rdata = THashLookupFromHash(set->hash, &lookup);
+    if (rdata) {
+        IPv4Type *found = rdata->data;
+        rrep.found = true;
+        rrep.json = found->json;
+        DatajsonUnlockData(rdata);
+        return rrep;
+    }
+    return rrep;
+}
+
+static DataJsonResultType DatajsonLookupIPv6(
+        Dataset *set, const uint8_t *data, const uint32_t data_len)
+{
+    DataJsonResultType rrep = { .found = false, .json = { .value = NULL, .len = 0 } };
+
+    if (set == NULL)
+        return rrep;
+
+    /* We can have IPv4 or IPV6 here due to ip.src and ip.dst implementation */
+    if (data_len != SC_IPV6_LEN && data_len != SC_IPV4_LEN)
+        return rrep;
+
+    IPv6Type lookup = { .json.value = NULL, .json.len = 0 };
+    memcpy(lookup.ipv6, data, data_len);
+    THashData *rdata = THashLookupFromHash(set->hash, &lookup);
+    if (rdata) {
+        IPv6Type *found = rdata->data;
+        rrep.found = true;
+        rrep.json = found->json;
+        DatajsonUnlockData(rdata);
+        return rrep;
+    }
+    return rrep;
+}
+
+DataJsonResultType DatajsonLookup(Dataset *set, const uint8_t *data, const uint32_t data_len)
+{
+    DataJsonResultType rrep = { .found = false, .json = { .value = 0 } };
+    if (set == NULL)
+        return rrep;
+
+    switch (set->type) {
+        case DATASET_TYPE_STRING:
+            return DatajsonLookupString(set, data, data_len);
+        case DATASET_TYPE_MD5:
+            return DatajsonLookupMd5(set, data, data_len);
+        case DATASET_TYPE_SHA256:
+            return DatajsonLookupSha256(set, data, data_len);
+        case DATASET_TYPE_IPV4:
+            return DatajsonLookupIPv4(set, data, data_len);
+        case DATASET_TYPE_IPV6:
+            return DatajsonLookupIPv6(set, data, data_len);
+        default:
+            break;
+    }
+    return rrep;
+}
+
+typedef int (*DatajsonOpFunc)(
+        Dataset *set, const uint8_t *data, const uint32_t data_len, const DataJsonType *json);
+
+static int DatajsonOpSerialized(Dataset *set, const char *string, const char *json,
```

Comment:
it remains unclear to me why we need so much code duplication from datasets... why can't it be refactored to reuse things from there?